src/bpfbytecode.cpp

```diff
@@ -83,51 +80,115 @@ BpfProgram &BpfBytecode::getProgramForProbe(const Probe &probe)
-  load_progs(resources.watchpoint_probes, btf, feature);
+  prepare_progs(resources.probes, btf, feature);
+  prepare_progs(resources.special_probes, btf, feature);
+  prepare_progs(resources.watchpoint_probes, btf, feature);
+
+  bpf_object__load(bpf_object_.get());
+
+  for (auto &[name, prog] : programs_) {
+    if (bt_debug.find(DebugStage::Verifier) != bt_debug.end()) {
+      std::cout << "BPF verifier log for " << name << ":\n";
+      std::cout << "--------------------------------------\n";
+      std::cout << prog.log_buf() << std::endl;
+    }
+
+    if (prog.fd() >= 0)
+      continue;
+
+    std::string log(prog.log_buf());
+    maybe_throw_helper_verifier_error(log,
+                                      "helper call is not allowed in probe",
+                                      " not allowed in probe");
+
+    std::stringstream errmsg;
+    errmsg << "Error loading BPF program for " << name;
+    if (bt_verbose) {
+      std::cerr << std::endl
+                << "Kernel error log: " << std::endl
+                << log << std::endl;
+      if (errno == ENOSPC) {
+        errmsg << "Error: Failed to load program, verification log buffer "
+               << "not big enough, try increasing the BPFTRACE_LOG_SIZE "
+               << "environment variable beyond the current value of "
+               << log_size_ << " bytes";
+      }
+    } else {
+      errmsg << " Use -v for full kernel error log.";
+    }
+    throw FatalUserException(errmsg.str());
+  }
 }
 
-void BpfBytecode::load_progs(const std::vector<Probe> &probes,
-                             BTF &btf,
-                             BPFfeature &feature)
+void BpfBytecode::prepare_progs(const std::vector<Probe> &probes,
+                                BTF &btf,
+                                BPFfeature &feature)
 {
   for (auto &probe : probes) {
     auto &program = getProgramForProbe(probe);
-    program.assemble(*this);
-    program.load(probe, *this, btf, feature);
+    program.set_prog_type(probe, feature);
+    program.set_expected_attach_type(probe, feature);
+    program.set_attach_target(probe, btf);
+    program.set_no_autoattach();
   }
 }
 
-bool BpfBytecode::create_maps()
+bool BpfBytecode::all_progs_loaded()
 {
-  int failed_maps = 0;
-  for (auto &map : maps_) {
-    BPFTRACE_LIBBPF_OPTS(bpf_map_create_opts, opts);
-    map.second.fd = bpf_map_create(static_cast<::bpf_map_type>(
-                                       map.second.type()),
-                                   map.second.bpf_name().c_str(),
-                                   map.second.key_size(),
-                                   map.second.value_size(),
-                                   map.second.max_entries(),
-                                   &opts);
-    if (map.second.fd < 0) {
-      failed_maps++;
-      LOG(ERROR) << "failed to create map: '" << map.first
-                 << "': " << strerror(errno);
-    }
-  }
-
-  if (failed_maps > 0) {
```

Comment:
How do we handle failing to create a map now?

Reply:
Good point, we don't. Thinking about it led me to realize that we now have fewer options to figure out what went wrong in prog loading or map creation in general. I tried to rework it to give user as many hints as possible, see `BpfBytecode::load_progs` for details. We can also discuss during office hours.